### PR TITLE
[ios][video] Background mode playback fix

### PR DESCRIPTION
**Pull request by hromovp**
# Why

When using background mode in iOS, app behavior is not necessary logical. e.g

1. When using `player.staysActiveInBackground = true` and `allowsPictureInPicture={false}` for `VideoView`:
    - upon locking screen with app in foreground **video stops**;
    - upon moving app to the background **video stops**;
2. When using `player.staysActiveInBackground = true` and `allowsPictureInPicture={true}` for `VideoView`:
    - upon moving app to the background PiP mode turns on and **video is playing**;
    - when locking screen while PiP mode is on - **audio is playing**;
    - upon locking screen with app in foreground **video stops**;

Considering there is more elegant way to handle background audio playback, configuration could be updated.

# How

- If using `player.staysActiveInBackground = true` set `audiovisualBackgroundPlaybackPolicy = .continuesIfPossible`
- Remove setTracksEnabled function

# Test Plan
1. In example video app, when using `player.staysActiveInBackground = true`  and `allowsPictureInPicture={false}` for `VideoView`:
    - upon locking screen with app in foreground **audio is playing** without PiP;
    - upon moving app to the background **audio is playing** without PiP;
2. When using `player.staysActiveInBackground = true` and `allowsPictureInPicture={true}` for `VideoView`:
    - upon moving app to the background PiP mode turns on and **video is playing**;
    - when locking screen while PiP mode is on - **video is playing**;
    - upon locking screen with app in foreground **audio is playing**;

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

### CODE REVIEW

**Comment by behenate** [comment]
Hi! 
I'd like to ask you some more questions about this part 
>```
>When using player.staysActiveInBackground = true and allowsPictureInPicture={false} for VideoView:
>. upon locking screen with app in foreground video stops;
>. upon moving app to the background video stops;
>When using player.staysActiveInBackground = true and allowsPictureInPicture={true} for VideoView:
>. upon moving app to the background PiP mode turns on and video is playing;
>. when locking screen while PiP mode is on - audio is playing;
>. upon locking screen with app in foreground video stops;
>```

I couldn't reproduce this behaviour at all. Behaviour in my testing was following:

1. When using `player.staysActiveInBackground = true` and `allowsPictureInPicture={false}` for VideoView:
    - upon locking screen with app in foreground video  **keeps playing**;
    - upon moving app to the background video **keeps playing**;
2. When using player.staysActiveInBackground = true and allowsPictureInPicture={true} for VideoView:
    - upon moving app to the background PiP mode turns on and video is playing;
    - when locking screen while PiP mode is on - audio is playing;
    - upon locking screen with app in foreground video **keeps playing**;

So basically i observed a correct behaviour.

**Comment by hromovp** [comment]
> Hi! I'd like to ask you some more questions about this part
> 
> > ```
> > When using player.staysActiveInBackground = true and allowsPictureInPicture={false} for VideoView:
> > . upon locking screen with app in foreground video stops;
> > . upon moving app to the background video stops;
> > When using player.staysActiveInBackground = true and allowsPictureInPicture={true} for VideoView:
> > . upon moving app to the background PiP mode turns on and video is playing;
> > . when locking screen while PiP mode is on - audio is playing;
> > . upon locking screen with app in foreground video stops;
> > ```
> 
> I couldn't reproduce this behaviour at all. Behaviour in my testing was following:
> 
> 1. When using `player.staysActiveInBackground = true` and `allowsPictureInPicture={false}` for VideoView:
>    
>    * upon locking screen with app in foreground video  **keeps playing**;
>    * upon moving app to the background video **keeps playing**;
> 2. When using player.staysActiveInBackground = true and allowsPictureInPicture={true} for VideoView:
>    
>    * upon moving app to the background PiP mode turns on and video is playing;
>    * when locking screen while PiP mode is on - audio is playing;
>    * upon locking screen with app in foreground video **keeps playing**;
> 
> So basically i observed a correct behaviour. How did you do your testing?

Oh, sorry. I am not using new architecture so that must be the reason.
Also the other reason might be that I am running HLS stream

**Comment by behenate** [comment]
The old architecture shouldn't be the problem. Expo-video was initially created and tested on old architecture and the background playback functionality was working.
Is the config plugin configured in your project? https://docs.expo.dev/versions/latest/sdk/video/#configurable-properties

**Comment by hromovp** [comment]
> The old architecture shouldn't be the problem. Expo-video was initially created and tested on old architecture and the background playback functionality was working. Is the config plugin configured in your project? https://docs.expo.dev/versions/latest/sdk/video/#configurable-properties

I use bare RN project, so it is not configured per se, but as far as  I understood the only thing that is needed is to add `UIBackgroundModes` - `audio`, and I did that.

My setup:
"react-native": "0.76.5",
"expo": "52.0.19",
"expo-video": "2.0.2",

Tested on: Iphone 12 mini, iOS 18.1.1

**Comment by behenate** [comment]
@hromovp Interesting. I'll do some more testing in a fresh project to reproduce this

**Comment by joeporpeglia** [comment]
@behenate I started experiencing this issue when switching from MP4 to HLS video. I created a corresponding MRE and issue here https://github.com/expo/expo/issues/34170

**Comment by henrikkoivisto** [comment]
Having same issue! @behenate @joeporpeglia .

**Comment by epzee** [comment]
@behenate fyi we're having the same issue. HLS video stops playing when app enters background mode on ios
```
    "expo": "~52.0.46",
    "expo-video": "~2.0.6",
    "react-native": "0.76.9",
```

File: packages/expo-video/ios/VideoManager.swift [hunk 1]
```diff
@@ -40,19 +40,15 @@ class VideoManager {
...
   func onAppBackgrounded() {
     for videoView in videoViews.allObjects {
       guard let player = videoView.player else {
         continue
       }
       if player.staysActiveInBackground == true {
-        player.setTracksEnabled(videoView.isInPictureInPicture)
+        player.pointer.audiovisualBackgroundPlaybackPolicy = .continuesIfPossible
       } else if !videoView.isInPictureInPicture {
         player.pointer.pause()
```
Comment (behenate) [inline packages/expo-video/ios/VideoManager.swift]: Let's set the audiovisualBackgroundPlaybackPolicy to `.pauses` when we don't need it. I assume keeping it in continuesIfPossible may have a negative effect on battery life.
```suggestion
        player.pointer.audiovisualBackgroundPlaybackPolicy = .continuesIfPossible
      } else if !videoView.isInPictureInPicture {
        player.ref.audiovisualBackgroundPlaybackPolicy = .pauses
        player.pointer.pause()
```

**Review by behenate — CHANGES_REQUESTED** [review]
@hromovp Hi, I'm really sorry for such a late review. Do you have the time to rebase the PR to remove the conflicts and apply my suggestion, or would you prefer me to take over? I'd like to have this merged by monday 😄

File: packages/expo-video/ios/VideoManager.swift (hunk 1 above)
Comment (hromovp) [reply packages/expo-video/ios/VideoManager.swift]: Hi @behenate, could you please check if this will work?

**Review by behenate — APPROVED** [review]
I tested it and it looks like it works well!
